src/Plugin/Block/PageHeaderBlock.php

```diff
@@ -0,0 +1,230 @@
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Routing\CurrentRouteMatch;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\node\Entity\Node;
+use Drupal\taxonomy\Entity\Term;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Drupal\localgov_core\Event\PageHeaderDisplayEvent;
+
+/**
+ * Provides a 'PageHeaderBlock' block.
+ *
+ * @Block(
+ *  id = "localgov_page_header_block",
+ *  admin_label = @Translation("Page header block"),
+ * )
+ */
+class PageHeaderBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * Core current_route_match service.
+   *
+   * @var \Drupal\Core\Routing\CurrentRouteMatch
+   */
+  protected $currentRouteMatch;
+
+  /**
+   * Core event_dispatcher service.
+   *
+   * @var \Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher
+   */
+  protected $eventDispatcher;
+
+  /**
+   * Core request_stack service.
+   *
+   * @var \Symfony\Component\HttpFoundation\RequestStack
+   */
+  protected $requestStack;
+
+  /**
+   * Core title_resolver service.
+   *
+   * @var \Drupal\Core\Controller\TitleResolver
+   */
+  protected $titleResolver;
+
+  /**
+   * Entity associated with the current route.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface|null
+   */
+  protected $entity = NULL;
+
+  /**
+   * The page title override.
+   *
+   * @var array|string|null
+   */
+  protected $title;
+
+  /**
+   * The page lede override.
+   *
+   * @var array|string|null
+   */
+  protected $lede;
+
+  /**
+   * Should the page header block be displayed?
+   *
+   * @var bool
+   */
+  protected $visible;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('current_route_match'),
+      $container->get('event_dispatcher'),
+      $container->get('request_stack'),
+      $container->get('title_resolver')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, CurrentRouteMatch $current_route_match, ContainerAwareEventDispatcher $event_dispatcher, RequestStack $request_stack, TitleResolver $title_resolver) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->currentRouteMatch = $current_route_match;
+    $this->eventDispatcher = $event_dispatcher;
+    $this->requestStack = $request_stack;
+    $this->titleResolver = $title_resolver;
+
+    // Find the entity, if any, associated with the current route.
+    // Adapted from https://www.computerminds.co.uk/drupal-code/get-entity-route
+    if (($route = $this->currentRouteMatch->getRouteObject()) && ($parameters = $route->getOption('parameters'))) {
```

Comment:
I need to think about this more generally, and look at the title resolver service, and what the options might be for template preprocess or something. I'm sure you've thought them through. Just a double think. But anyway... while reading this code. I was thinking.

Why getting the raw parameters and checking for the keys and so forth?

foreach ($route_match->getParameters() as $paramater) { if ($parameter instanceof EntityInterface && ...)

Also why does it need a canonical link template?

Sorry just getting my head round it.

Reply:
So this is the danger of just copying and pasting code (although I did include a link to where I found it).

Digging into the Drupal source code; this is largely copied from the content translation module: https://git.drupalcode.org/project/drupal/-/blob/8.9.x/core/modules/content_translation/content_translation.module#L662. 

Git blame points to the main logic for this here: https://www.drupal.org/project/drupal/issues/2303525#comment-9032709 - which doesn't give any reasons as to why it's done in the way it is.

We can certainly remove the canonical link template as that's only needed for translated content: https://www.drupal.org/project/drupal/issues/2479377

Reply:
> Why getting the raw parameters and checking for the keys and so forth?
> 
> foreach ($route_match->getParameters() as $paramater) { if ($parameter instanceof EntityInterface && ...)

This seems the easiest way of determining the whether the route references an entity. `$this->currentRouteMatch->getParameters()` is also an option, but it returns more parameters and checking whether any of them is an entity looks to be more involved.

> Also why does it need a canonical link template?

It doesn't. Removed